internal/bguploader.go

```diff
@@ -1,146 +1,171 @@
 package internal
 
 import (
+	"context"
 	"io/ioutil"
+	"os"
 	"path/filepath"
 	"strings"
-	"sync"
 	"sync/atomic"
 	"time"
 
-	"github.com/spf13/viper"
 	"github.com/wal-g/tracelog"
+	"golang.org/x/sync/semaphore"
 )
 
 // BgUploader represents the state of concurrent WAL upload
 type BgUploader struct {
 	// pg_[wals|xlog]
 	dir string
 
-	// count of running goroutines
-	parallelWorkers int32
-
-	// usually defined by WALG_DOWNLOAD_CONCURRENCY
-	maxParallelWorkers int32
-
-	// waitgroup to handle Stop gracefully
-	running sync.WaitGroup
-
-	// every file is attempted only once
-	started map[string]interface{}
-
 	// uploading structure
 	uploader *WalUploader
 
-	// to control amount of work done in one cycle of archive_command
-	totalUploaded int32
+	preventWalOverwrite bool
 
-	mutex sync.Mutex
+	// ctx signals internals to keep/stop enqueueing more uploads
+	ctx context.Context
+	// cancelFunc signals internals to stop enqueuing more uploads
+	cancelFunc context.CancelFunc
 
-	preventWalOverwrite bool
+	// workerCount tracks number of concurrent uploaders. Limited to maxParallelWorkers
+	workerCount *semaphore.Weighted
+	// maxParallelWorkers usually defined by WALG_DOWNLOAD_CONCURRENCY
+	maxParallelWorkers int32
```

Comment:
Explain what maxParallelWorkers does?  Looks like it's the number of concurrent goroutines to upload WAL files.